copy.go

```diff
@@ -259,12 +265,85 @@ func copyGraph(ctx context.Context, src content.ReadOnlyStorage, dst content.Sto
 		if exists {
 			return copyNode(ctx, proxy.Cache, dst, desc, opts)
 		}
-		return copyNode(ctx, src, dst, desc, opts)
+		return mountOrCopyNode(ctx, src, dst, desc, opts)
 	}
 
 	return syncutil.Go(ctx, limiter, fn, root)
 }
 
+// mountOrCopyNode tries to mount the node, if not falls back to copying.
+func mountOrCopyNode(ctx context.Context, src content.ReadOnlyStorage, dst content.Storage, desc ocispec.Descriptor, opts CopyGraphOptions) error {
+	// Need MountFrom and it must be a blob
+	if opts.MountFrom == nil || descriptor.IsManifest(desc) {
+		return copyNode(ctx, src, dst, desc, opts)
+	}
+
+	mounter, ok := dst.(registry.Mounter)
+	if !ok {
+		// mounting is not supported by the destination
+		return copyNode(ctx, src, dst, desc, opts)
+	}
+
+	sourceRepositories, err := opts.MountFrom(ctx, desc)
+	if err != nil {
+		// Technically this error is not fatal, we can still attempt to copy the node
+		// But for consistency with the other callbacks we bail out.
+		return err
+	}
+
+	if len(sourceRepositories) == 0 {
+		return copyNode(ctx, src, dst, desc, opts)
+	}
+
+	skipContent := errors.New("skip content")
+	for i, sourceRepository := range sourceRepositories {
+		// try mounting this source repository
+		var mountFailed bool
+		getContent := func() (io.ReadCloser, error) {
+			// the invocation of getContent indicates that mounting has failed
+			mountFailed = true
+
+			if len(sourceRepositories)-1 == i {
+				// this is the last iteration so we need to actually get the content and do the copy
+				// but first call the PreCopy function
+				if opts.PreCopy != nil {
+					if err := opts.PreCopy(ctx, desc); err != nil {
+						return nil, err
+					}
+				}
+				return src.Fetch(ctx, desc)
+			}
+
+			// We want to return an error that we will test for from mounter.Mount()
+			return nil, skipContent
```

Comment:
nit: Would something like this be more readable?

```suggestion
			if i < len(sourceRepositories)-1 {
				// If this is not the last one, skip this source and try next one
				// We want to return an error that we will test for from mounter.Mount()
				return nil, skipContent
			}

			// this is the last iteration so we need to actually get the content and do the copy
			// but first call the PreCopy function
			if opts.PreCopy != nil {
				if err := opts.PreCopy(ctx, desc); err != nil {
					return nil, err
				}
			}
			return src.Fetch(ctx, desc)
```

Reply:
I like that better as well.